mono/metadata/icall.c

```diff
@@ -1347,69 +1346,64 @@ ves_icall_System_ValueType_InternalGetHashCode (MonoObject *this_obj, MonoArray
 			continue;
 		if (mono_field_is_deleted (field))
 			continue;
+		gpointer addr = (guint8*)MONO_HANDLE_RAW (this_obj)  + field->offset;
 		/* FIXME: Add more types */
 		switch (field->type->type) {
 		case MONO_TYPE_I4:
-			result ^= *(gint32*)((guint8*)this_obj + field->offset);
+			result ^= *(gint32*)addr;
 			break;
 		case MONO_TYPE_PTR:
-			result ^= mono_aligned_addr_hash (*(gpointer*)((guint8*)this_obj + field->offset));
+			result ^= mono_aligned_addr_hash (*(gpointer*)addr);
 			break;
 		case MONO_TYPE_STRING: {
 			MonoString *s;
-			s = *(MonoString**)((guint8*)this_obj + field->offset);
+			s = *(MonoString**)addr;
 			if (s != NULL)
 				result ^= mono_string_hash_internal (s);
 			break;
 		}
 		default:
 			if (!values)
-				values = g_newa (MonoObject*, mono_class_num_fields (klass));
-			o = mono_field_get_value_object_checked (mono_object_domain (this_obj), field, this_obj, error);
-			if (!is_ok (error)) {
-				mono_error_set_pending_exception (error);
-				return 0;
-			}
-			values [count++] = o;
+				values = g_newa (MonoObjectHandle, mono_class_num_fields (klass));
```

Comment:
The number of allocated handles is equal to the number of fields in a user-provided class. We try not to do that.

It's not clear to me why we need to get the fields ahead of time.  Can't we just use this first loop to count and assign the fields to the array in the second loop, below, without an intermediate stack allocated array?